core/trino-main/src/main/java/io/trino/execution/SqlStage.java

```diff
@@ -289,7 +289,8 @@ private synchronized void updateFinalTaskInfo(TaskInfo finalTaskInfo)
 
     private synchronized void checkAllTaskFinal()
     {
-        if (stateMachine.getState().isDone() && tasksWithFinalInfo.containsAll(tasks.keySet())) {
+        if (stateMachine.getState().isDone() && (
```

Comment:
Why not just use `tasksWithFinalInfo.size() == tasks.size()` like we're doing in `updateTaskStatus` now? Seems like there is no way for the key sets to contain different `TaskId` values without it being a bug.